pkg/controller/infrastructureconfig/infrastructureconfig_controller.go

```diff
@@ -0,0 +1,105 @@
+	return add(mgr, newReconciler(mgr, status, c))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, status *statusmanager.StatusManager, c cnoclient.Client) reconcile.Reconciler {
+	return &ReconcileInfrastructureConfig{client: c, scheme: mgr.GetScheme(), status: status, apiAndIngressVIPsSyncer: &apiAndIngressVipsSynchronizer{}}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("infrastructureconfig-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource config.openshift.io/v1/Infrastructure
+	err = c.Watch(&source.Kind{Type: &configv1.Infrastructure{}}, &handler.EnqueueRequestForObject{}, onPremPlatformPredicate(), platformStatusChangePredicate())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileInfrastructureConfig{}
+
+// ReconcileInfrastructureConfig reconciles a cluster Infrastructure object
+type ReconcileInfrastructureConfig struct {
+	client                  cnoclient.Client
+	scheme                  *runtime.Scheme
+	status                  *statusmanager.StatusManager
+	apiAndIngressVIPsSyncer vipsSynchronizer
+}
+
+// Reconcile watches Infrastructure.config.openshift.io/cluster and syncs the
+// new and deprecated API & Ingress VIP fields to have consistent APIs between
+// versions.
+func (r *ReconcileInfrastructureConfig) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("Reconciling Infrastructure.config.openshift.io %s\n", request.Name)
+
+	// Only check on the default infrastructure config
+	if request.Name != names.INFRASTRUCTURE_CONFIG {
+		log.Printf("Ignoring Infrastructure without default name " + names.INFRASTRUCTURE_CONFIG)
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch the infrastructure config
+	infraConfig := &configv1.Infrastructure{}
+	err := r.client.Default().CRClient().Get(ctx, request.NamespacedName, infraConfig)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Return and don't requeue
+			log.Println("Object seems to have been deleted")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		err = fmt.Errorf("Error while reading infrastructures.%s/cluster: %w", configv1.GroupName, err)
+		log.Println(err)
+		return reconcile.Result{}, err
+	}
+
+	// Sync API & Ingress VIPs
+	updatedInfraConfig := r.apiAndIngressVIPsSyncer.VipsSynchronize(infraConfig)
+
+	if !reflect.DeepEqual(updatedInfraConfig.Status, infraConfig.Status) {
+		err = r.client.Default().CRClient().Status().Update(ctx, updatedInfraConfig, &client.UpdateOptions{})
```

Comment:
This will work better as a server-side-apply to avoid clobbering unknown fields as schemas evolve: see https://github.com/openshift/cluster-etcd-operator/pull/833/files#r941800222 as an example.